x-pack/plugins/spaces/server/new_platform/spaces_service/spaces_service.ts

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { first } from 'rxjs/operators';
+import { Observable, Subscription } from 'rxjs';
+import { Legacy } from 'kibana';
+import { SavedObjectsService } from 'src/legacy/server/kbn_server';
+import {
+  Logger,
+  ElasticsearchServiceSetup,
+  HttpServiceSetup,
+  KibanaRequest,
+} from 'src/core/server';
+import { DEFAULT_SPACE_ID } from '../../../common/constants';
+import { SecurityPlugin } from '../../../../security';
+import { SpacesClient } from '../../lib/spaces_client';
+import { getSpaceIdFromPath } from '../../lib/spaces_url_parser';
+import { SpacesConfigType } from '../config';
+
+type RequestFacade = KibanaRequest | Legacy.Request;
+
+export interface SpacesServiceSetup {
+  scopedClient(request: RequestFacade): SpacesClient;
+
+  getSpaceId(request: RequestFacade): string;
+
+  isInDefaultSpace(request: RequestFacade): boolean;
+}
+
+interface CacheEntry {
+  spaceId: string;
+  isInDefaultSpace: boolean;
+}
+
+interface SpacesServiceDeps {
+  http: HttpServiceSetup;
+  elasticsearch: ElasticsearchServiceSetup;
+  savedObjects: SavedObjectsService;
+  getSecurity: () => SecurityPlugin | undefined;
+  config$: Observable<SpacesConfigType>;
+  spacesAuditLogger: any;
+}
+
+export class SpacesService {
+  private configSubscription$?: Subscription;
+
+  private readonly contextCache: WeakMap<any, CacheEntry> = new WeakMap();
+
+  constructor(private readonly log: Logger, private readonly serverBasePath: string) {
+    this.contextCache = new WeakMap();
+  }
+
+  public async setup({
+    http,
+    elasticsearch,
+    savedObjects,
+    getSecurity,
+    config$,
+    spacesAuditLogger,
+  }: SpacesServiceDeps): Promise<SpacesServiceSetup> {
+    let config: SpacesConfigType = await config$.pipe(first()).toPromise();
+
+    this.configSubscription$ = config$.subscribe({
+      next: updatedConfig => {
+        config = updatedConfig;
+      },
+    });
+
+    const adminClient = await elasticsearch.adminClient$.pipe(first()).toPromise();
```

Comment:
`first` and not seeing any updated version seems problematic here as well.

Reply:
I used `first` here based on my understanding that the clients cannot be changed once created: https://github.com/elastic/kibana/pull/35429#discussion_r279355338.

Is this the wrong way to go about it?

Reply:
technically, as they exposed as Observables, we shouldn't relay on the first value only. 
Is this limitation not to re-create elasticsearch clients with updated config is going to stay? @azasypkin @elastic/kibana-platform

Reply:
> Is this limitation not to re-create elasticsearch clients with updated config is going to stay? 

IMO allowing elasticsearch config to change without requiring admins to restart Kibana server would be beneficial (e.g. to turn on `elasticsearch.logQueries`) and we may want to support that at some point. I'm not sure if we should account for that now if it's not going to happen anytime soon though (likely not before we entirely deprecate and remove legacy elasticsearch plugin) and would be interested in hearing what is the current platform team plan.

Reply:
In the meantime, should we change the exposed type from `Observable` to `BehaviorSubject`?

Reply:
By deciding to go with Observables for all data we fundamentally decided to build a reactive architecture. So I think the Core API "contract" requires consumers of the API to handle underlying updates even if Core doesn't yet make use of it. And in fleshing out the API we rather want to have the optionality to e.g. change elasticsearch config in the future than being restricted.